samples/nrf_desktop/src/modules/config.c

```diff
@@ -0,0 +1,221 @@
+
+static int settings_set(int argc, char **argv, void *val_ctx)
+{
+	if (argc != 1) {
+		return -ENOENT;
+	}
+
+	if (IS_ENABLED(CONFIG_DESKTOP_MOTION_OPTICAL_ENABLE)) {
+		char *end;
+		u8_t id = strtol(argv[0], &end, 10);
+		if (*end != '\0') {
+			LOG_WRN("Loaded config ID is not a valid number");
+			return 0;
+		}
+
+		if ((id == CONFIG_EVENT_ID_RESERVED) ||
+		    (id >= CONFIG_EVENT_ID_MAX)) {
+			LOG_WRN("Not supported config: %u", id);
+			return 0;
+		}
+
+		__ASSERT_NO_MSG(id > CONFIG_EVENT_ID_RESERVED);
+		__ASSERT_NO_MSG(id < CONFIG_EVENT_ID_MAX);
+
+		if (!loaded_data[id]) {
+			loaded_data[id] = k_malloc(REPORT_SIZE_USER_CONFIG - 1);
+
+			if (loaded_data[id] == NULL) {
+				LOG_ERR("No memory");
+				module_set_state(MODULE_STATE_ERROR);
+				return -ENOMEM;
+			}
+		}
+
+		int len = settings_val_read_cb(val_ctx,
+			loaded_data[id],
+			REPORT_SIZE_USER_CONFIG - 1);
+		if (len < 0) {
+			LOG_ERR("Can't read config (id:%u err:%d)", id, len);
+			return len;
+		}
+	}
+
+	return 0;
+}
+
+int commit(void)
+{
+	/* All settings loaded, including FCB duplicates.
+	 * We can apply values to application.
+	 */
+	for (u8_t id = 1; id < CONFIG_EVENT_ID_MAX; id++) {
+		if (loaded_data[id]) {
+
+			struct config_event *event = new_config_event();
+			event->id = id;
+			event->stored = true;
+			memcpy(event->data, loaded_data[id],
+			       sizeof(event->data));
+
+			EVENT_SUBMIT(event);
+
+			k_free(loaded_data[id]);
+		}
+	}
+
+	return 0;
+}
+
+static int update_config(const u8_t config_id, u8_t *data)
+{
+	int err = 0;
+
+	char key[KEY_LEN];
+	snprintk(key, sizeof(key), "%s/%u", MODULE_NAME, config_id);
+
+	LOG_INF("Store %s", log_strdup(key));
+	err = settings_save_one(key, data, REPORT_SIZE_USER_CONFIG - 1);
+
+	return err;
+}
+
+static int enable_settings(void)
+{
+	int err = 0;
+
+	if (IS_ENABLED(CONFIG_SETTINGS)) {
+		if (IS_ENABLED(CONFIG_DESKTOP_CONFIG_CHANNEL_ENABLE)) {
+			static struct settings_handler sh = {
+				.name = MODULE_NAME,
+				.h_set = settings_set,
+				.h_commit = commit,
+			};
+
+			err = settings_register(&sh);
+			if (err) {
+				LOG_ERR("Cannot register settings handler");
+				goto error;
+			}
+		}
+
+		err = settings_load();
+		if (err) {
+			LOG_ERR("Cannot load settings");
+			goto error;
+		}
+
+		LOG_INF("Settings loaded");
+	}
+
+error:
+	return err;
+}
+
+static void init(void)
+{
+	int err = enable_settings();
+
+	if (err) {
+		module_set_state(MODULE_STATE_ERROR);
+	} else {
+		module_set_state(MODULE_STATE_READY);
+	}
+}
+
+static bool event_handler(const struct event_header *eh)
+{
+	if (IS_ENABLED(CONFIG_DESKTOP_CONFIG_CHANNEL_ENABLE)) {
+		if (is_config_event(eh)) {
+			struct config_event *event = cast_config_event(eh);
+
+			if (!event->stored) {
+				int err = update_config(event->id, event->data);
+
+				if (err) {
+					LOG_ERR("Cannot store value, err %d",
+						err);
+				}
+			}
+
+			return false;
+		}
+	}
+
+	if (is_module_state_event(eh)) {
+		const void * const req_modules[] = {
```

Comment:
on dongle we will have no hids, bas, ble_adv nor motion_optical... which means setting will never be initialized

Reply:
the best way to fix it is to add dependency on main

Reply:
At this moment, dongle is not using this code. Scanning has its own handlers to settings. In my opinion, it should be integrated for dongle when we will add any configurable value there (connection interval?).

Reply:
discussed offline to integrate with dongle